src/expr/src/interpret.rs

```diff
@@ -941,30 +974,34 @@ impl Interpreter for Trace {
         left: Self::Summary,
         right: Self::Summary,
     ) -> Self::Summary {
-        // TODO: this is duplicative! If we have more than one or two special-cased functions
-        // we should find some way to share the list between `Trace` and `ColumnSpecs`.
         let (left_pushdownable, right_pushdownable) = match SpecialBinary::for_func(func) {
             None => func.is_monotone(),
             Some(special) => special.pushdownable,
         };
-        (left && left_pushdownable) || (right && right_pushdownable)
+        left.apply_fn(left_pushdownable)
+            .max(right.apply_fn(right_pushdownable))
     }
 
     fn variadic(&self, func: &VariadicFunc, exprs: Vec<Self::Summary>) -> Self::Summary {
         if !func.is_associative() && exprs.len() >= ColumnSpecs::MAX_EVAL_ARGS {
             // We can't efficiently evaluate functions with very large argument lists;
             // see the comment on ColumnSpecs::MAX_EVAL_ARGS for details.
-            return false;
+            return TraceSummary::Unknown;
         }
 
         let pushdownable_fn = func.is_monotone();
         exprs
             .into_iter()
-            .any(|pushdownable_arg| pushdownable_arg && pushdownable_fn)
+            .map(|pushdownable_arg| pushdownable_arg.apply_fn(pushdownable_fn))
+            .max()
+            .unwrap_or(TraceSummary::Constant)
     }
 
     fn cond(&self, cond: Self::Summary, then: Self::Summary, els: Self::Summary) -> Self::Summary {
-        cond || (then || els)
+        // We don't actually need to be able to predict the condition precisely to predict the output,
+        // since it's just a boolean.
```

Comment:
Is it because it's a boolean or because you can union the ranges of the two branches for a conservative estimate?

Reply:
The latter's probably the more obvious explanation... switched it over.